code/starfield/starfield.cpp

```diff
@@ -1791,23 +1770,18 @@ void stars_draw_debris()
 			} else {
 				alpha = 1.0f;
 			}
-			
-			vm_vec_add( &tmp, &d->last_pos, &Eye_position );
-			//g3_draw_laser( &d->pos,d->size,&tmp,d->size, TMAP_FLAG_TEXTURED|TMAP_FLAG_XPARENT, 25.0f );
-			material mat_params;
-			material_set_unlit(&mat_params, Debris_vclips[d->vclip].bm + frame, alpha, true, true);
-			g3_render_laser_2d(&mat_params, &d->pos, d->size, &tmp, d->size, 25.0f);
-		}
 
-		vm_vec_sub( &d->last_pos, &d->pos, &Eye_position );
+			// scale alpha from 0 at max range to full at 60% range
+			alpha *= (vdist - MAX_DIST_RANGE) / -(MAX_DIST_RANGE * 0.6f);
 
-		vdist = vm_vec_mag_quick(&d->last_pos);
+			g3_transfer_vertex(&pnt, &mdebris.pos);
 
-		if ( (vdist < MIN_DIST_RANGE) || (vdist > MAX_DIST_RANGE) )
-			d->active = 0;
+			batching_add_bitmap(Debris_vclips[mdebris.vclip].bm + frame, &pnt, 0, mdebris.size, alpha);
```

Comment:
I do not see a call to actually execute the batches you add here in this function and if I remember correctly that is what actually makes the GPU do stuff in that system. While that does not cause issues with the rendering of the debris it might change your measurements since no graphics operations actually take place within this scope but the cost is added somewhere else.

This is still a worthwhile change but might not be the x30 performance increase you are seeing in your measurements.

Reply:
On second thought, you are right, that number I remembered was specifically this loop itself because I wanted the gauge the effect of increasing the total number of debris (its been a little bit since I wrote this code). But in general I am aware that the batch renders themselves should also be counted, which did come out favorably although I can't recall the exact numbers produced.